File: fs/numbers/src/NumbersIterative.fs
namespace src

module NumbersIterative =

    let fibonacci n =
        let rec fibonacci_iter n i aux acc =
            match n with
            | n when n <= 1 -> n
            | n when i = n -> acc
            | _ -> fibonacci_iter n (i + 1) acc (acc + aux)
        fibonacci_iter n 2 1 1

    let factorial n = 
        let rec factorial_iter n acc =
            match n with
            | 1 -> acc
            | n -> factorial_iter (n - 1) (n * acc)
        factorial_iter n 1

    let sum_numbers n =
        let rec sum_numbers_iter n acc =
            match n with
            | 0 -> acc
            | n -> sum_numbers_iter (n - 1) (n + acc)
        sum_numbers_iter n 0

File: _experimental/_haskell/numbers/src/NumbersIterative.hs
module NumbersIterative (fibonacci, factorial, sumNumbers) where

fibonacci :: Int -> Int
fibonacci n = fib_iter n 0 1
  where
    fib_iter n acc2 acc1
        | n <= 0    = acc2
        | n <= 2    = acc1 + acc2
        | otherwise = fib_iter (n - 1) acc1 (acc1 + acc2)

factorial :: Int -> Int
factorial n = factorial_iter n 1
  where
    factorial_iter n acc
        | n <= 1    = acc
        | otherwise = factorial_iter (n - 1) (n * acc)

sumNumbers :: Int -> Int
sumNumbers n = sum_iter n 0
  where
    sum_iter n acc
        | n <= 0    = acc
        | otherwise = sum_iter (n - 1) (n + acc)